prov/verbs/src/verbs_domain.c

```diff
@@ -385,6 +385,44 @@ static void fi_ibv_domain_process_exp(struct fi_ibv_domain *domain)
 	domain->use_odp = fi_ibv_gl_data.use_odp;
 }
 
+static int
+fi_ibv_post_send_track_credits(struct ibv_qp *qp, struct ibv_send_wr *wr,
+			       struct ibv_send_wr **bad_wr)
+{
+	struct fi_ibv_cq *cq =
+		container_of(((struct fi_ibv_ep *)qp->qp_context)->util_ep.tx_cq,
+			     struct fi_ibv_cq, util_cq);
+	int credits = (int)ofi_atomic_dec32(&cq->credits);
+	int ret;
+
+	if (credits < 0) {
+		FI_DBG(&fi_ibv_prov, FI_LOG_EP_DATA, "CQ credits not available,"
+		       " retry later\n");
+		ofi_atomic_inc32(&cq->credits);
+		return ENOMEM;
```

Comment:
Is this the right return value?  Should this be negative or set errno instead?

Reply:
This tries to mimic ibv_post_send which returns ENOMEM on full send queue. `fi_ibv_handle_post` should convert a positive error value to appropriate libfabric error number.